src/components/Table.tsx

```diff
@@ -81,42 +88,93 @@ const renderField = (row: TableDataRow, column: TableColumn) => {
 	return value == null ? '' : value;
 };
 
-export default class Table extends React.Component<
-	TableProps,
+export default class Table<T> extends React.Component<
+	TableProps<T>,
 	{
-		sortColumn: null | string;
+		allChecked: boolean;
 		reverse: boolean;
+		checkedItems: T[];
+		sortColumn: null | keyof T;
 	}
 > {
-	constructor(props: TableProps) {
+	constructor(props: TableProps<T>) {
 		super(props);
 
+		if (props.onChecked && !props.rowKey) {
+			throw new Error(
+				'A `rowKey` property must be provided if using `onChecked` with a Table component',
+			);
+		}
+
 		this.state = {
-			sortColumn: null,
+			allChecked: false,
 			reverse: false,
+			checkedItems: [],
+			sortColumn: null,
 		};
 	}
 
-	sortData(data: TableDataRow[]): TableDataRow[] {
+	isChecked(item: T) {
+		const rowKey = this.props.rowKey;
+		if (!rowKey) {
+			return false;
+		}
+
+		const identifier = item[rowKey];
+		return !!find(this.state.checkedItems, x => x[rowKey] === identifier);
+	}
+
+	sortData(data: T[]): T[] {
 		if (this.state.sortColumn === null) {
 			return data;
 		}
 
-		const collection = (sortBy(data.slice(), item => {
-			const sortableValue = item[this.state.sortColumn!];
+		const collection = sortBy(data.slice(), item => {
+			const sortableValue = item[this.state.sortColumn!] as any;
 			return isPlainObject(sortableValue)
 				? (sortableValue as any).value
 				: sortableValue;
-		}) as any) as TableDataRow[];
+		});
 
 		if (this.state.reverse) {
 			reverse(collection);
 		}
 
-		return collection;
+		return (collection as any) as T[];
+	}
+
+	toggleAllChecked() {
+		const allChecked = !this.state.allChecked;
+		const checkedItems = allChecked ? this.props.data.slice() : [];
+
+		if (this.props.onChecked) {
+			this.props.onChecked(checkedItems);
+		}
+
+		this.setState({ allChecked, checkedItems });
+	}
+
+	toggleChecked(item: T) {
+		const rowKey = this.props.rowKey;
+		if (!rowKey) {
+			return false;
+		}
+
+		const identifier = item[rowKey];
+
+		const isChecked = !this.isChecked(item);
+		const checkedItems = isChecked
+			? this.state.checkedItems.concat(item)
+			: this.state.checkedItems.filter(x => x[rowKey] !== identifier);
+
+		if (this.props.onChecked) {
+			this.props.onChecked(checkedItems);
+		}
+
+		this.setState({ allChecked: false, checkedItems });
```

Comment:
I'm not sure about the `allChecked: false` part.
Shouldn't this become `true` if we select all the table items one by one?

Reply:
Yup, ill update it